src/mpi/datatype/typerep/src/typerep_yaksa_pack.c

```diff
@@ -499,235 +415,252 @@ int MPIR_Typerep_wait(MPIR_Typerep_req typerep_req)
-
-                if (src_ptr) {
-                    MPI_Aint unpacked_size;
-                    op_unpack((char *) source_buf +
-                              src_type_stride * accumulated_count, source_dtp_size,
-                              src_ptr, 1, source_dtp, 0, &unpacked_size, op);
-                    MPIR_Typerep_copy((char *) target_buf + MPIR_Ptr_to_aint(curr_loc), src_ptr,
-                                      type_size * count);
-                } else {
-                    max_pack_bytes = type_size * count;
-                    if (source_dtp_size == source_dtp_extent) {
-                        mpi_errno =
-                            op_pack((char *) source_buf +
-                                    src_type_stride * accumulated_count, count,
-                                    type, 0,
-                                    (char *) target_buf +
-                                    MPIR_Ptr_to_aint(curr_loc), max_pack_bytes,
-                                    &actual_pack_bytes, op);
-                    } else {
-                        char *ptr = MPL_malloc(source_dtp_extent, MPL_MEM_OTHER);
-                        MPI_Aint unpacked_size;
-                        mpi_errno =
-                            op_pack((char *) source_buf +
-                                    src_type_stride * accumulated_count, count,
-                                    type, 0, ptr, max_pack_bytes, &actual_pack_bytes, op);
-                        MPIR_Typerep_unpack(ptr, source_dtp_size,
-                                            (char *) target_buf + MPIR_Ptr_to_aint(curr_loc), 1,
-                                            source_dtp, 0, &unpacked_size);
-                        MPL_free(ptr);
-                    }
-                }
-                if (curr_len % type_size == 0) {
-                    i++;
-                    if (i != vec_len) {
-                        curr_loc = typerep_vec[i].iov_base;
-                        curr_len = typerep_vec[i].iov_len;
-                    }
-                } else {
-                    curr_loc = (void *) ((char *) curr_loc + type_extent * count);
-                    curr_len -= type_size * count;
-                }
+    bool use_yaksa = MPIR_Typerep_reduce_is_supported(op, source_dtp);
+    int source_is_contig, target_is_contig;
+    MPIR_Datatype_is_contig(source_dtp, &source_is_contig);
+    MPIR_Datatype_is_contig(source_dtp, &target_is_contig);
 
-                accumulated_count += count;
-            }
+    if (use_yaksa) {
+        if (source_is_packed || source_is_contig) {
+            mpi_errno = typerep_op_unpack(source_buf, target_buf, target_count, target_dtp, op);
+        } else if (target_is_contig) {
+            mpi_errno = typerep_op_pack(source_buf, target_buf, source_count, source_dtp, op);
+        } else {
+            MPI_Aint source_dtp_size;
+            MPIR_Datatype_get_size_macro(source_dtp, source_dtp_size);
```

Comment:
nitpick: we have `MPIR_Pack_size` to calculate the buffer size here and a few other places

Reply:
For using `MPIR_Pack_size`, do we need to check if the buffer is already packed or not?

Reply:
When landed in this branch, the source buffer is not packed.